core/src/main/scala/kafka/server/Kip500Broker.scala

```diff
@@ -325,11 +327,22 @@ class Kip500Broker(val config: KafkaConfig,
           1, s"${SocketServer.ControlPlaneMetricPrefix}RequestHandlerAvgIdlePercent", SocketServer.ControlPlaneThreadPrefix)
       }
 
+      // Block until we've caught up on the metadata log
+      brokerMetadataListener.initiallyCaughtUpFuture.get()
+      // Start log manager, which will perform (potentially lengthy) recovery-from-unclean-shutdown if required.
+      logManager.startup()
+      // Start other services that we've delayed starting, in the appropriate order.
```

Comment:
LGTM overall. 

I'm wondering if we can capture this dependency in a future chain after the `logManager.startup` is complete. That way, it'd be more clear that there is a dependency on the log manager and the broker registration here. That should also allow the `BrokerLifecycleManager` to return a future on startup and have the `Kip500Broker` chain these startups after successful registration outside the `BrokerLifecycleManager` instead of the ugly await inside.

Reply:
Yeah, I was thinking about this type of possibility as well.  I think we're going to want to make `BrokerMetadataListener.start()` return a future, so that PR would be a good place to consider something like this.